opnode/l1/receipts.go

```diff
@@ -0,0 +1,85 @@
+package l1
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/trie"
+
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const MaxReceiptRetry = 3
+
+func fetchReceipts(ctx context.Context, log log.Logger, receiptHash common.Hash, txs types.Transactions, getBatch batchCallContextFn) (types.Receipts, error) {
+	if len(txs) == 0 {
+		if receiptHash != types.EmptyRootHash {
+			return nil, fmt.Errorf("no transactions, but got non-empty receipt trie root: %s", receiptHash)
+		}
+		return nil, nil
+	}
+
+	receipts := make([]*types.Receipt, len(txs))
+	receiptRequests := make([]rpc.BatchElem, len(txs))
+	for i := 0; i < len(txs); i++ {
+		receipts[i] = new(types.Receipt)
+		receiptRequests[i] = rpc.BatchElem{
```

Comment:
I would imagine there are some limits in geth or at infura/alchemy that impose a limit on the number of batch elements in an rpc request. For example, our `proxyd` allows for only 100 per RPC request, see https://github.com/ethereum-optimism/optimism/blob/3b82da3e886dad9a1bf9ed441b032f192ce9bed9/go/proxyd/server.go#L27

Reply:
Good catch, I'll refactor it to split it into multiple batch requests

Reply:
Just looked at geth a bit, I think that geth will only block on size of request/response and duration of request processing, I didn't see any code that restricted based on size of batch but I didn't test anything. Unsure what other clients do and unsure if infra providers do anything special at their proxy layer

Reply:
Implemented `fetchBatched`, with edge-case tests for splitting/retries/parallel/errors behavior. It's a drop-in replacement for the existing batch-call function type. During testing of the source we just mock the outer single batch calls, but we use the tested splitting&parallelizing implementation of it in E2E.